File: android/app/src/main/java/com/antnagi/nagisheart/engine/StoryEngine.kt
package com.antnagi.nagisheart.engine

import com.antnagi.nagisheart.data.*

sealed class NodeResolution {
    data class Found(
        val nodeId: String,
        val node: StoryNode,
        val visual: SceneVisual?
    ) : NodeResolution()

    data class EndingReached(
        val endingId: String,
        val definition: EndingDefinition
    ) : NodeResolution()

    data class NotFound(val id: String, val reason: String) : NodeResolution()
}

class StoryEngine(
    private val nodes: Map<String, StoryNode>,
    private val flow: FlowData,
    private val routers: Map<String, Router>,
    private val sceneVisuals: Map<String, SceneVisual>,
    private val endings: EndingsData,
    private val conditionParser: ConditionParser = ConditionParser()
) {

    private val routerIds: Set<String> = routers.keys
    private val nodeIds: Set<String> = nodes.keys

    fun isRouter(id: String): Boolean = id in routerIds
    fun isNode(id: String): Boolean = id in nodeIds

    /**
     * Resolve an ID to a presentable node.
     * If the ID is a router, evaluates rules and follows the chain
     * until a story node or ending is reached.
     * Routers and nodes are explicitly distinguished — no null fallback.
     */
    fun resolve(id: String, state: GameState): NodeResolution {
        var currentId = id
        val visited = mutableSetOf<String>()

        while (currentId !in visited) {
            visited.add(currentId)

            when {
                isRouter(currentId) -> {
                    val router = routers[currentId]!!
                    val result = resolveRouter(router, state)
                    state.applyEffects(result.sideEffects)
                    currentId = result.targetId
                }

                isNode(currentId) -> {
                    return NodeResolution.Found(
                        nodeId = currentId,
                        node = nodes[currentId]!!,
                        visual = sceneVisuals[currentId]
                    )
                }

                isEndingNode(currentId) -> {
                    val key = currentId.removePrefix("end_")
                    val def = endings.definitions[key]
                        ?: return NodeResolution.NotFound(currentId, "Ending definition not found")
                    return NodeResolution.EndingReached(currentId, def)
                }

                else -> {
                    return NodeResolution.NotFound(currentId, "ID is neither a node nor a router")
                }
            }
        }

        return NodeResolution.NotFound(id, "Router loop detected")
    }

    /**
     * Get the next node ID from flow after the current node.
     * Checks byRoute overrides first (matching current path/mj), then default.
     */
    fun getNextNodeId(currentId: String, state: GameState): String? {
        for ((routeName, overrides) in flow.byRoute) {
            if (currentId in overrides) {
                val match = when (routeName) {
                    "M" -> state.getString("mj") == "M"
                    "J" -> state.getString("mj") == "J"
                    "dream" -> state.getString("path") == "dream"
                    "stay" -> state.getString("path") == "stay"
                    "bad" -> state.getString("path") == "bad"
                    else -> false
                }
                if (match) return overrides[currentId]
            }
        }
        return flow.default[currentId]
    }

    /**
     * Filter choices to only those whose condition (if any) is satisfied.
     */
    fun getVisibleChoices(choices: List<Choice>, state: GameState): List<Choice> {
        return choices.filter { choice ->
            choice.condition?.let { conditionParser.evaluate(it, state) } ?: true
        }
    }

    /**
     * Determine the next destination after a choice is made.
     * Returns the target ID from transition, or null if flow should be used.
     */
    fun getEndingDefinitions(): Map<String, EndingDefinition> = endings.definitions

    fun processChoiceTransition(choice: Choice): String? {
        return when (choice.transition?.type) {
            "goto" -> choice.transition.target
            "ending" -> choice.transition.target ?: "ending_resolver"
            else -> null
        }
    }

    // --- Private helpers ---

    private data class RouterResult(
        val targetId: String,
        val sideEffects: List<Effect>
    )

    private fun resolveRouter(router: Router, state: GameState): RouterResult {
        for (rule in router.rules) {
            if (conditionParser.evaluate(rule.condition, state)) {
                return RouterResult(rule.target, rule.sideEffects)
            }
        }
        return RouterResult(router.fallback, router.fallbackSideEffects)
    }

    private fun isEndingNode(id: String): Boolean {
        if (!id.startsWith("end_")) return false
        val key = id.removePrefix("end_")
        return key in endings.definitions
    }
}

File: web/src/engine/StoryEngine.js
import { ConditionParser } from './ConditionParser.js';

export class StoryEngine {
  constructor({ nodes, flow, routers, sceneVisuals, endings }) {
    this._nodes = nodes;
    this._flow = flow;
    this._routers = routers;
    this._sceneVisuals = sceneVisuals;
    this._endings = endings;
    this._conditionParser = new ConditionParser();
    this._routerIds = new Set(Object.keys(routers));
    this._nodeIds = new Set(Object.keys(nodes));
  }

  isRouter(id) { return this._routerIds.has(id); }
  isNode(id) { return this._nodeIds.has(id); }

  resolve(id, state) {
    let currentId = id;
    const visited = new Set();

    while (!visited.has(currentId)) {
      visited.add(currentId);

      if (this.isRouter(currentId)) {
        const router = this._routers[currentId];
        const result = this._resolveRouter(router, state);
        state.applyEffects(result.sideEffects);
        currentId = result.targetId;
      } else if (this.isNode(currentId)) {
        return {
          type: 'found',
          nodeId: currentId,
          node: this._nodes[currentId],
          visual: this._sceneVisuals[currentId] || null
        };
      } else if (this._isEndingNode(currentId)) {
        const key = currentId.replace(/^end_/, '');
        const def = this._endings.definitions[key];
        if (!def) return { type: 'notFound', id: currentId, reason: 'Ending definition not found' };
        return { type: 'endingReached', endingId: currentId, definition: def };
      } else {
        return { type: 'notFound', id: currentId, reason: 'ID is neither a node nor a router' };
      }
    }

    return { type: 'notFound', id, reason: 'Router loop detected' };
  }

  getNextNodeId(currentId, state) {
    const byRoute = this._flow.byRoute || {};
    for (const [routeName, overrides] of Object.entries(byRoute)) {
      if (currentId in overrides) {
        let match = false;
        switch (routeName) {
          case 'M':     match = state.getString('mj') === 'M'; break;
          case 'J':     match = state.getString('mj') === 'J'; break;
          case 'dream': match = state.getString('path') === 'dream'; break;
          case 'stay':  match = state.getString('path') === 'stay'; break;
          case 'bad':   match = state.getString('path') === 'bad'; break;
        }
        if (match) return overrides[currentId];
      }
    }
    return this._flow.default[currentId] || null;
  }

  getVisibleChoices(choices, state) {
    if (!choices) return [];
    return choices.filter(c =>
      c.condition ? this._conditionParser.evaluate(c.condition, state) : true
    );
  }

  processChoiceTransition(choice) {
    if (!choice.transition) return null;
    switch (choice.transition.type) {
      case 'goto':   return choice.transition.target;
      case 'ending': return choice.transition.target || 'ending_resolver';
      default:       return null;
    }
  }

  getEndingDefinitions() {
    return this._endings.definitions;
  }

  // Resolve the ending_resolver router against current state → 'end_*' node id.
  // Used when the story is fast-forwarded (section-skip) past its final content.
  resolveEndingId(state) {
    const router = this._routers['ending_resolver'];
    if (!router) return null;
    const { targetId } = this._resolveRouter(router, state);
    return targetId || null;
  }

  _resolveRouter(router, state) {
    for (const rule of router.rules) {
      if (this._conditionParser.evaluate(rule.condition, state)) {
        return { targetId: rule.target, sideEffects: rule.sideEffects || [] };
      }
    }
    return { targetId: router.fallback, sideEffects: router.fallbackSideEffects || [] };
  }

  _isEndingNode(id) {
    if (!id.startsWith('end_')) return false;
    const key = id.replace(/^end_/, '');
    return key in this._endings.definitions;
  }
}
